crowbar_framework/app/models/updater_service.rb

```diff
@@ -55,15 +63,27 @@ def create_proposal
 
   def apply_role_pre_chef_call(old_role, role, all_nodes)
     @logger.debug("Updater apply_role_post_chef_call: entering #{all_nodes.inspect}")
-    # Remove [:updater][:one_shot_run] flag from node
+    # Remove [:updater][...] flags from node
     all_nodes.each do |n|
       node = NodeObject.find_node_by_name n
       unless node[:updater].nil?
-        node[:updater][:one_shot_run] = false
-        @logger.debug("Updater apply_role_post_chef_call: delete [:updater][:one_shot_run] for #{node.name} (#{node[:updater].inspect}")
-        node.save
+        if node.role? "updater"
+          node[:updater][:one_shot_run] = false
+          @logger.debug("Updater apply_role_post_chef_call: delete [:updater][:one_shot_run] for #{node.name} (#{node[:updater].inspect}")
+          node.save
+        end
+        if node.role? "updater-crowbar-upgrade"
+          node[:updater][:upgrade_one_shot_run] = false
+          @logger.debug("Updater apply_role_post_chef_call: delete [:updater][:upgrade_one_shot_run] for #{node.name} (#{node[:updater].inspect}")
+
+          # By modifying the runlist we ensure that only upgrade related recipes are executed
+          node.run_list.run_list_items.clear
+          node.run_list.run_list_items << "role[updater-crowbar-upgrade]"
+          node.save
```

Comment:
As discussed on IRC I think  it should be possible to leverage ```node.set_state``` here with just a few tweaks to the updater template.

This would handle the state transition and update of the run_list in a much cleaner way IMO